File: li2/li2/plp/imperative2/exemplos/Exemplo2.java
package li2.plp.imperative2.exemplos;

import li2.plp.expressions1.util.TipoPrimitivo;
import li2.plp.expressions2.expression.*;
import li2.plp.imperative1.command.*;
import li2.plp.imperative1.declaration.*;
import li2.plp.imperative1.memory.ListaValor;
import li2.plp.imperative2.Programa;
import li2.plp.imperative2.command.*;
import li2.plp.imperative2.declaration.*;
import li2.plp.imperative2.memory.*;

/**
 * Atividade 2.
 *
 * Ilustra o escopo dinamico da li2, executando:
 *
 *   { var x = 0, proc p (int y) {x := x + y};
 *     { var x = 1; call p(3); write(x) };
 *     call p(4); write(x) }
 *
 * Como o escopo e dinamico, a primeira chamada (dentro do bloco interno onde
 * x = 1) imprime 4; na segunda chamada apenas o x externo (= 0) esta em
 * escopo e o valor produzido tambem e 4.  Saida esperada: 4 4.
 */
public class Exemplo2 {

	public static void main(String[] args) throws Exception {

		// ---------- Declaracoes do bloco externo ----------
		// var x = 0
		Declaracao declaracaoVariavel =
				new DeclaracaoVariavel(new Id("x"), new ValorInteiro(0));

		// proc p (int y) { x := x + y }
		ListaDeclaracaoParametro parametros =
				new ListaDeclaracaoParametro(
						new DeclaracaoParametro(new Id("y"), TipoPrimitivo.INTEIRO));
		DefProcedimento defP = new DefProcedimento(
				parametros,
				new Atribuicao(new Id("x"),
						new ExpSoma(new Id("x"), new Id("y"))));
		Declaracao declaracaoProcedimento =
				new DeclaracaoProcedimento(new Id("p"), defP);

		Declaracao declaracoes =
				new DeclaracaoComposta(declaracaoVariavel, declaracaoProcedimento);

		// ---------- Bloco interno: { var x = 1; call p(3); write(x) } ----------
		Comando blocoInterno = new ComandoDeclaracao(
				new DeclaracaoVariavel(new Id("x"), new ValorInteiro(1)),
				new SequenciaComando(
						new ChamadaProcedimento(new Id("p"),
								new ListaExpressao(new ValorInteiro(3))),
						new Write(new Id("x"))));

		// ---------- Corpo do bloco externo ----------
		// blocoInterno ; call p(4) ; write(x)
		Comando corpo = new SequenciaComando(
				blocoInterno,
				new SequenciaComando(
						new ChamadaProcedimento(new Id("p"),
								new ListaExpressao(new ValorInteiro(4))),
						new Write(new Id("x"))));

		Comando programa = new ComandoDeclaracao(declaracoes, corpo);

		AmbienteExecucaoImperativa2 ambiente =
				new ContextoExecucaoImperativa2(new ListaValor());

		ListaValor saida = new Programa(programa).executar(ambiente);
		System.out.println("Exemplo2 -> saida: " + saida);  // esperado: 4 4
	}
}

File: python/li2/plp/imperative2/Exemplo2.py
"""
Atividade 2 (Lab 05 - li2).

Executa:
  { var x = 0, proc p (int y) {x := x + y};
    { var x = 1; call p(3); write(x) };
    call p(4); write(x)
  }
"""
from li2.plp.expressions1.util.TipoPrimitivo import TipoPrimitivo
from li2.plp.expressions2.expression.ExpSoma import ExpSoma
from li2.plp.expressions2.expression.Id import Id
from li2.plp.expressions2.expression.ValorInteiro import ValorInteiro
from li2.plp.imperative1.command.Atribuicao import Atribuicao
from li2.plp.imperative1.command.ComandoDeclaracao import ComandoDeclaracao
from li2.plp.imperative1.command.SequenciaComando import SequenciaComando
from li2.plp.imperative1.command.Write import Write
from li2.plp.imperative1.declaration.DeclaracaoComposta import DeclaracaoComposta
from li2.plp.imperative1.declaration.DeclaracaoVariavel import DeclaracaoVariavel
from li2.plp.imperative1.memory.ContextoCompilacaoImperativa import ContextoCompilacaoImperativa
from li2.plp.imperative1.memory.ListaValor import ListaValor
from li2.plp.imperative2.command.ChamadaProcedimento import ChamadaProcedimento
from li2.plp.imperative2.command.ListaExpressao import ListaExpressao
from li2.plp.imperative2.declaration.DeclaracaoParametro import DeclaracaoParametro
from li2.plp.imperative2.declaration.DeclaracaoProcedimento import DeclaracaoProcedimento
from li2.plp.imperative2.declaration.DefProcedimento import DefProcedimento
from li2.plp.imperative2.declaration.ListaDeclaracaoParametro import ListaDeclaracaoParametro
from li2.plp.imperative2.memory.ContextoExecucaoImperativa2 import ContextoExecucaoImperativa2
from li2.plp.imperative2.Programa import Programa


class Exemplo2:

    @staticmethod
    def main():
        x = Id("x")
        y = Id("y")
        p = Id("p")

        def_p = DefProcedimento(
            ListaDeclaracaoParametro(DeclaracaoParametro(y, TipoPrimitivo.INTEIRO)),
            Atribuicao(x, ExpSoma(x, y)),
        )

        bloco_interno = ComandoDeclaracao(
            DeclaracaoVariavel(x, ValorInteiro(1)),
            SequenciaComando(ChamadaProcedimento(p, ListaExpressao(ValorInteiro(3))), Write(x)),
        )

        declaracoes = DeclaracaoComposta(DeclaracaoVariavel(x, ValorInteiro(0)), DeclaracaoProcedimento(p, def_p))
        comando = ComandoDeclaracao(
            declaracoes,
            SequenciaComando(
                bloco_interno,
                SequenciaComando(ChamadaProcedimento(p, ListaExpressao(ValorInteiro(4))), Write(x)),
            ),
        )
        programa = Programa(comando)

        if programa.checaTipo(ContextoCompilacaoImperativa(ListaValor())):
            saida = programa.executar(ContextoExecucaoImperativa2(ListaValor()))
            print(saida)
        else:
            print("Comando mal tipado.")


if __name__ == "__main__":
    Exemplo2.main()
